components/style/gecko_selector_impl.rs

```diff
@@ -12,104 +12,130 @@ use stylesheets::Stylesheet;
-pub enum AnonBoxPseudoElement {
-    MozText,
-    MozOtherNonElement,
-    MozAnonymousBlock,
-    MozAnonymousPositionedBlock,
-    MozMathMLAnonymousBlock,
-    MozXULAnonymousBlock,
-
-    MozHorizontalFramesetBorder,
-    MozVerticalFramesetBorder,
-    MozLineFrame,
-    MozButtonContent,
-    MozButtonLabel,
-    MozCellContent,
-    MozDropdownList,
-    MozFieldsetContent,
-    MozFramesetBlank,
-    MozDisplayComboboxControlFrame,
-
-    MozHTMLCanvasContent,
-    MozInlineTable,
-    MozTable,
-    MozTableCell,
-    MozTableColumnGroup,
-    MozTableColumn,
-    MozTableWrapper,
-    MozTableRowGroup,
-    MozTableRow,
-
-    MozCanvas,
-    MozPageBreak,
-    MozPage,
-    MozPageContent,
-    MozPageSequence,
-    MozScrolledContent,
-    MozScrolledCanvas,
-    MozScrolledPageSequence,
-    MozColumnContent,
-    MozViewport,
-    MozViewportScroll,
-    MozAnonymousFlexItem,
-    MozAnonymousGridItem,
-
-    MozRuby,
-    MozRubyBase,
-    MozRubyBaseContainer,
-    MozRubyText,
-    MozRubyTextContainer,
-
-    MozTreeColumn,
-    MozTreeRow,
-    MozTreeSeparator,
-    MozTreeCell,
-    MozTreeIndentation,
-    MozTreeLine,
-    MozTreeTwisty,
-    MozTreeImage,
-    MozTreeCellText,
-    MozTreeCheckbox,
-    MozTreeProgressMeter,
-    MozTreeDropFeedback,
-
-    MozSVGMarkerAnonChild,
-    MozSVGOuterSVGAnonChild,
-    MozSVGForeignContent,
-    MozSVGText,
+// TODO: automatize this list from
+// http://searchfox.org/mozilla-central/source/layout/style/nsCSSPseudoElements.h and
+impl PseudoElement {
+    #[inline]
+    fn before() -> Self {
+        PseudoElement(atom!(":before"), false)
+    }
+
+    #[inline]
+    fn after() -> Self {
+        PseudoElement(atom!(":before"), false)
+    }
+
+    #[inline]
+    fn as_atom(&self) -> &Atom {
+        &self.0
+    }
+
+    #[inline]
+    fn is_anon_box(&self) -> bool {
+        self.1
+    }
+
+    #[inline]
+    fn from_atom_unchecked(atom: Atom, is_anon_box: bool) -> Self {
+        if cfg!(debug_assertions) {
+            match Self::from_atom(&*atom, true) {
+                Some(pseudo) => {
+                    assert_eq!(pseudo.is_anon_box(), is_anon_box);
+                    return pseudo;
+                }
+                None => panic!("Unknown pseudo: {:?}", atom),
+            }
+        }
+
+        PseudoElement(atom, is_anon_box)
+    }
+
+    #[inline]
+    fn from_atom(atom: &WeakAtom, in_ua: bool) -> Option<Self> {
+        use std::slice::Iter;
+        let mut slice = atom.as_slice();
+
+        let len = slice.len();
+        let mut iter = slice.iter();
+
+        if iter.next().map(|c| *c) != Some(':' as u16) {
+            return None
+        }
+
+        // This type is just needed because a plain iter.map() doesn't derive
+        // Clone because the closure is not Clone.
+        #[derive(Clone)]
+        struct Utf16toASCIIIter<'a>(Iter<'a, u16>);
+
+        impl<'a> Iterator for Utf16toASCIIIter<'a> {
```

Comment:
(Would still prefer to not introduce the nasty stuff at all though, even for a temporary patch)